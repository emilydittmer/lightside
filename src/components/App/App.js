import React, { Component } from "react";
import { Route } from "react-router-dom";
import { fetchAPI } from "../../utils/apiCalls";
import Loader from "../Loader/Loader";
import Showcase from "../Showcase/Showcase";
import Form from "../Form/Form";
import Catalog from "../Catalog/Catalog";
import "./App.scss";

class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoading: true,
      error: "",
      category: "",
      films: [],
      people: [],
      planets: [],
      vehicles: [],
      favorites: []
    };
  }

  async componentDidMount() {
    const data = await fetchAPI();
    const { films, people, planets, vehicles } = data;
    this.setState({ films, people, planets, vehicles, isLoading: false });
  }

  toggleFavorite = (category, name) => {
    let clonedState = this.state[category].slice();
    let index = clonedState.findIndex(obj => {
      return obj.name === name;
    });
    clonedState[index].favorited = !clonedState[index].favorited;
    this.setState({ [category]: clonedState });
    if (clonedState[index].favorited) {
      this.addFavorite(clonedState[index]);
    } else {
      this.removeFavorite(clonedState[index].name);
    }
  };

  addFavorite = obj => {
    console.log("adding favorite");
    this.setState({ favorites: [...this.state.favorites, obj] });
  };

  removeFavorite = name => {
    console.log("removing favorite");
    const { favorites } = this.state;
    const filteredFaves = favorites.filter(obj => {
      return obj.name !== name;
    });
    this.setState({ favorites: filteredFaves });
  };

  render() {
    const { films, people, planets, vehicles, favorites } = this.state;

    return (
      <div className="App">
        {this.state.isLoading ? (
          <Loader />
        ) : (
          <Showcase key={"films"} films={films} />
        )}
        <Form counter={favorites.length} />
        <Route
          exact
          path="/"
          render={() => <Catalog key={"home"} counter={1} data={[]} />}
        />
        <Route
          exact
          path="/people"
          render={() => (
            <Catalog
              toggle={this.toggleFavorite}
              key={"people"}
              counter={people.length}
              data={people}
            />
          )}
        />
        <Route
          exact
          path="/planets"
          render={() => (
            <Catalog
              toggle={this.toggleFavorite}
              key={"planets"}
              counter={planets.length}
              data={planets}
            />
          )}
        />
        <Route
          exact
          path="/vehicles"
          render={() => (
            <Catalog
              toggle={this.toggleFavorite}
              key={"vehicles"}
              counter={vehicles.length}
              data={vehicles}
            />
          )}
        />
        <Route
          exact
          path="/favorites"
          render={() => (
            <Catalog
              toggle={this.toggleFavorite}
              key={"favorites"}
              counter={favorites.length}
              data={favorites}
            />
          )}
        />
      </div>
    );
  }
}

export default App;
