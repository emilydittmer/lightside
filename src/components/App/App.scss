@import '../../sass/_mixins.scss';
@import '../../sass/_variables.scss';

.App {
  display: grid;
  grid-template-rows: 10% 30% 15% 45%;
  background: url(../../images/background.jpg);
  background-repeat: no-repeat;
}
header {
  grid-row: 1/2;
  h1 {
    font-family: 'Starjedi';
    font-size: 5em;
    text-align: center;
    margin: 0;
    color:hsla(51, 100%, 50%, 1);
  }
}

.loading {
  text-align: center;
  margin-top: 20px;
  height: 100vh;
  .loading-text {
  color:hsla(51, 100%, 50%, 1);
  font-family: 'Starjedi';
  }
}