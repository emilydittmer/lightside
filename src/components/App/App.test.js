import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import mockData from '../../data/mock-data';

describe('App', () => {

  let fetchData;
  let data;
  let wrapper;

  beforeEach(() => {
    //
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('calls fetch with the right data', () => {
    //
  });

  it('should update state with the resolved fetch data', () => {
    //
  });

  it('should be able to toggle favorited objects', () => {
    //
  });

  it('should add objects to favorites', () => {
    //
  });

  it('should removed objects from favorites', () => {
    //
  });
})
