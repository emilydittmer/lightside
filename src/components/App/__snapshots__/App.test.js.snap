// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should add objects to favorites 1`] = `ShallowWrapper {}`;

exports[`App should match snapshot 1`] = `ShallowWrapper {}`;

exports[`App should remove objects from favorites 1`] = `ShallowWrapper {}`;

exports[`App should removed objects from favorites 1`] = `ShallowWrapper {}`;
